lib/auth/join_tpm_test.go

```diff
@@ -0,0 +1,344 @@
+			JoinMethod: types.JoinMethodTPM,
+			Roles:      []types.SystemRole{types.RoleNode},
+			TPM: &types.ProvisionTokenSpecV2TPM{
+				Allow: []*types.ProvisionTokenSpecV2TPM_Rule{
+					{
+						Description:  "ekpub only",
+						EKPublicHash: "mock-ekpub-hashed",
+					},
+					{
+						Description:         "ekcert only",
+						EKCertificateSerial: "mock-ekcert-serial",
+					},
+					{
+						Description:         "both",
+						EKPublicHash:        "mock-ekpub-hashed-1",
+						EKCertificateSerial: "mock-ekcert-serial-1",
+					},
+				},
+			},
+		}
+		if mutate != nil {
+			mutate(&spec)
+		}
+		return spec
+	}
+	joinRequest := func() *types.RegisterUsingTokenRequest {
+		return &types.RegisterUsingTokenRequest{
+			HostID:       "host-id",
+			Role:         types.RoleNode,
+			PublicTLSKey: tlsPublicKey,
+			PublicSSHKey: sshPublicKey,
+		}
+	}
+
+	caPool := x509.NewCertPool()
+	require.True(t, caPool.AppendCertsFromPEM([]byte(apifixtures.TLSCACertPEM)))
+
+	allowRulesNotMatched := require.ErrorAssertionFunc(func(t require.TestingT, err error, i ...interface{}) {
+		require.ErrorContains(t, err, "validated tpm attributes did not match any allow rules")
+		require.True(t, trace.IsAccessDenied(err))
+	})
+	tests := []struct {
+		name          string
+		setEnterprise bool
+
+		tokenSpec types.ProvisionTokenSpecV2
+
+		validateReturnTPM *tpm.ValidatedTPM
+		validateReturnErr error
+
+		initReq    *proto.RegisterUsingTPMMethodInitialRequest
+		wantParams tpm.ValidateParams
+
+		assertError require.ErrorAssertionFunc
+	}{
+		{
+			name:          "success, ekpub",
+			setEnterprise: true,
+			assertError:   require.NoError,
+
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				JoinRequest: joinRequest(),
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkKey{
+					EkKey: []byte("mock-ekpub"),
+				},
+				AttestationParams: attParams,
+			},
+			tokenSpec: tokenSpec(nil),
+			wantParams: tpm.ValidateParams{
+				EKKey:        []byte("mock-ekpub"),
+				AttestParams: tpm.AttestationParametersFromProto(attParams),
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKPubHash: "mock-ekpub-hashed",
+			},
+		},
+		{
+			name:          "success, ekcert",
+			setEnterprise: true,
+			assertError:   require.NoError,
+
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				JoinRequest: joinRequest(),
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkCert{
+					EkCert: []byte("mock-ekcert"),
+				},
+				AttestationParams: attParams,
+			},
+			tokenSpec: tokenSpec(func(v2 *types.ProvisionTokenSpecV2) {
+				v2.TPM.EKCertAllowedCAs = []string{apifixtures.TLSCACertPEM}
+			}),
+			wantParams: tpm.ValidateParams{
+				EKCert:       []byte("mock-ekcert"),
+				AttestParams: tpm.AttestationParametersFromProto(attParams),
+				AllowedCAs:   caPool,
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKCertSerial:   "mock-ekcert-serial",
+				EKCertVerified: true,
+			},
+		},
+		{
+			name:          "success, both",
```

Comment:
What makes this different from "success, ekcert" ?

Reply:
I'll improve the description, the idea is that the rule is checking that both the pub hash and cert serial match.